Adds Globalnet doc (#243)
submariner-io/submariner: docs/globalnet.md

# Introduction

Submariner is a tool built to connect overlay networks of different Kubernetes clusters. These clusters can be on different public clouds or on-premise. An important use case for Submariner is to connect disparate independent clusters into a single cohesive multi-cluster.
 However, one key limitation of submariner's current design is that it doesn't support overlapping CIDRs (ServiceCIDR and ClusterCIDR) across these clusters. Each cluster must use distinct CIDRs that don't conflict or overlap with any other cluster that is going to be part of cluster fleet.

<!-- TODO: Add diagram showing the problem --> 

This is largely problematic because most actual deployments use the default CIDRs for a cluster so every cluster ends up using the same CIDRs. Changing CIDRs on existing clusters is a very disruptive process and requires a cluster restart. So submariner needs a way to allow clusters with overlapping CIDRs to connect together.

# Architecture

> To support overlapping CIDRs in clusters connected through submariner, we're introducing a new component called Global Private Network, GlobalNet (`globalnet`). This GlobalNet is a virtual network specifically to support submariner's multi-cluster solution with a Global CIDR. Each cluster is given a subnet from this Global Private Network, configured as new cluster parameter `GlobalCIDR` (e.g. 169.254.0.0/16) which is configurable at time of deployment

Reviewer: ```suggestion
To support overlapping CIDRs in clusters connected through submariner, we're introducing a new component called Global Private Network or GlobalNet (`globalnet`). GlobalNet is a virtual network specifically to support submariner's multi-cluster solution using a Global CIDR. Each cluster is given a subnet from the Global Private Network, configured as new cluster parameter `GlobalCIDR` (e.g. 169.254.0.0/16) which is configurable at time of deployment
```

> Once configured, each service and pod that requires cross-cluster access is allocated an IP, `GlobalIp` from this `GlobalCIDR` and it is annotated on the Pod/Service object. This `GlobalIp` is used for all cross-cluster communication to and from to this Pod/Service. Routing and IPTable/OVS/OVN rules are configured to use this IP for ingress and egress. All address translations occur on the Gateway node.

Reviewer: ```suggestion
Once configured, each service and pod that requires cross-cluster access is allocated an IP, `GlobalIp`, from the `GlobalCIDR` which is then annotated on the Pod/Service object. This `GlobalIp` is used for all cross-cluster communication to and from to this Pod/Service. Routing and IPTable/OVS/OVN rules are configured to use this IP for ingress and egress. All address translations occur on the Gateway node.
```

## submariner-globalnet

> Submariner GlobalNet is a coponent that provides cross-cluster connectivity between pods and services using their GlobalIps. Compiled as binary `submariner-globalnet`, it is the controller that is responsible for maintaining a pool of global IPs, allocating IPs from the globalIp pool to pods and services, annotating services and pods with their global Ip, and configuring the required rules on gateway node to provide cross-cluster connectivity using GlobalIps. It mainly consists of two key components: the IP Address Manager and Globalnet.

Reviewer: ```suggestion
Submariner GlobalNet is a coponent that provides cross-cluster connectivity between pods and services using their GlobalIps. Compiled as binary `submariner-globalnet`, it is the controller that is responsible for maintaining a pool of global IPs, allocating IPs from this pool to pods and services and annotating them, and configuring the required rules on the gateway node to provide cross-cluster connectivity using the GlobalIps. It mainly consists of two key components: the IP Address Manager and Globalnet.
```

# Introduction

> Submariner is a tool build to connect overlay networks of different Kubernetes clusters. These clusters can be on different public clouds or OnPremises. An important use case for Submariner is to connect disparage independent clusters into a single cohesive multi-cluster.

Reviewer: ```suggestion
Submariner is a tool built to connect overlay networks of different Kubernetes clusters. These clusters can be on different public clouds or on-premise. An important use case for Submariner is to connect disparate independent clusters into a single cohesive multi-cluster.
```

>  However, one key limitation of submariner's current design is that it doesn't support overlapping CIDRs across these clusters. Each cluster must use a distinct ClusterCIDR that doesn't conflict or overlap with any other cluster that is going to be part of multicluster.

Reviewer: ```suggestion
 However, one key limitation of submariner's current design is that it doesn't support overlapping CIDRs across these clusters. Each cluster must use a distinct ClusterCIDR that doesn't conflict or overlap with any other cluster that is going to be part of the cluster fleet.
```

<!-- TODO: Add diagram showing the problem --> 

> This is largely problematic coz most actual deployments use the default CIDR for cluster, this end up using the same CIDR. Changing CIDR on existing clusters is a very disruptive process and requires a cluster restart. So we require a means to allow clusters with Overlapping CIDRs connect together with submariner.

Reviewer: ```suggestion
This is largely problematic because most actual deployments use the default CIDRs for a cluster so every cluster ends up using the same CIDRs. Changing CIDRs on existing clusters is a very disruptive process and requires a cluster restart. So submariner needs a way to allow clusters with overlapping CIDRs to connect together..
```

# Architecture

> To support overlapping CIDRs in clusters connected through submariner, we're introducing a new component called Global Private Network, or GlobalNet (`globalnet`). This GlobalNet is a virtual network with scope as submariner multi-cluster and CIDR 169.254.0.0/16. Each cluster is given a subnet from this Global Private Network, configured as new cluster parameter `GlobalCIDR`.

Reviewer 2: Not sure if we want to capture the CIDR (i.e., 169.254.0.0/16) in the document as this will eventually be coming from the Operator/Subctl during deployment.

> To support overlapping CIDRs in clusters connected through submariner, we're introducing a new component called Global Private Network, or GlobalNet (`globalnet`). This GlobalNet is a virtual network with scope as submariner multi-cluster and CIDR 169.254.0.0/16. Each cluster is given a subnet from this Global Private Network, configured as new cluster parameter `GlobalCIDR`.

Reviewer: ```suggestion
To support overlapping CIDRs in clusters connected through submariner, we're introducing a new component called Global Private Network, or GlobalNet (`globalnet`). This GlobalNet is a virtual network specifically to support  submariner's multi-cluster solution with CIDR 169.254.0.0/16. Each cluster is given a subnet from this Global Private Network, configured as a new cluster parameter `GlobalCIDR`.
```

> 

Reviewer 2: nit: translations occur at Gateway node -> translations occur on the Gateway node

> Once configured, each service and pod that requires cross cluster access is allocated an IP, `GlobalIp` from this `GlobalCIDR` and it is annotated to the POD/Service object. This `GlobalIp` is used for all cross-cluster communication from and to this POD/Service. Routing and IPTable/OVS/OVN rules are configured to use this IP for Ingress/Egress. All address translations occur at Gateway node.

Reviewer 2: nit: cross cluster -> cross-cluster

Reviewer 2: there are multiple such occurrences in the doc.

> Once configured, each service and pod that requires cross cluster access is allocated an IP, `GlobalIp` from this `GlobalCIDR` and it is annotated to the POD/Service object. This `GlobalIp` is used for all cross-cluster communication from and to this POD/Service. Routing and IPTable/OVS/OVN rules are configured to use this IP for Ingress/Egress. All address translations occur at Gateway node.

Reviewer: ```suggestion
Once configured, each service and pod that requires cross-cluster access is allocated an IP, `GlobalIp`, from this `GlobalCIDR` and it is annotated on the Pod/Service object. This `GlobalIp` is used for all cross-cluster communication to and from the Pod/Service. Routing and IPTable/OVS/OVN rules are configured to use this IP for ingress and egress. All address translations occur on the Gateway node.
```

> This is achieved by a new controller, Submariner Globalnet.

Reviewer: This sentence can be removed as it is redundant with the following section.

## submariner-globalnet

> Submariner GlobalNet provides cross cluster connectivity between pods and services using GlobalIp. Compiled as binary `submariner-globalnet`, it is the controller that is responsible for maintaining and allocating pool of globalIPs, annotating services and pods with their globalIp, and finally configuring the required rules on gateway node to provide cross cluster connectivity using GlobalIp. It mainly consists of two key components.

Reviewer 2: nit: maintaining and "allocating pool of globalIPs" -> this might give an impression that globalnet allocates a bunch of globalIPs. Shall we modify the sentence to "maintaining the GlobalPool and allocating an IP from this pool to Pods/Services, annotating them accordingly.....

Feel free to ignore this comment if you think my interpretation is wrong.

> Submariner GlobalNet provides cross cluster connectivity between pods and services using GlobalIp. Compiled as binary `submariner-globalnet`, it is the controller that is responsible for maintaining and allocating pool of globalIPs, annotating services and pods with their globalIp, and finally configuring the required rules on gateway node to provide cross cluster connectivity using GlobalIp. It mainly consists of two key components.

Reviewer: ```suggestion
Submariner GlobalNet is a controller component that provides cross-cluster connectivity between pods and services using their GlobalIps. Compiled as binary `submariner-globalnet`, it is responsible for maintaining and allocating a pool of global IPs, annotating services and pods with their GlobalIp, and configuring the required rules on the gateway node to provide cross cluster connectivity using the GlobalIps. It mainly consists of two key components: the IP Address Manager and Globalnet.
```

> ### IP Address Manager

Reviewer: ```suggestion
### IP Address Manager (IPAM)
```

> First key component of Globalnet controller is IP Address Manager (IPAM). This module does the following:

Reviewer: ```suggestion
The IP Address Manager component does the following:
```

> * Create a pool of IP addresses based on `GlobalCIDR` configured on cluster.

Reviewer: ```suggestion
* Creates a pool of IP addresses based on the `GlobalCIDR` configured for the cluster.
```

> * Whenever a Pod/Service is created, allocate a GlobalIp from the GlobalIp pool.

Reviewer 2: It would be good to clarify that this IP is "virtual" and is not physically assigned to the POD interface

Author: We specify it is virtual in the introduction.

> * Whenever a Pod/Service is created, allocate a GlobalIp from the GlobalIp pool.

Reviewer: ```suggestion
* On creation of a Pod/Service, allocates a GlobalIp from the GlobalIp pool.
```

> * Annotate the Pod/Service with the `submariner.io/GlobalIp=<global-ip>`.

Reviewer: ```suggestion
* Annotates the Pod/Service with `submariner.io/GlobalIp=<global-ip>`.
```

> * Whenever Pod/Service is deleted, release the IP allocated from the GlobalIp pool.

Reviewer: ```suggestion
* On deletion of a Pod/Service, releases its GlobalIp back to the pool.
```

### Globalnet

> This is the core component and is responsible for programming the routing entries and IPTable rules. This module does the following:

Reviewer: ```suggestion
This component is responsible for programming the routing entries and IPTable rules and does the following:
```

* Creates initial IPTables chains for Globalnet rules.

> * Whenever a pod is annotated with Global IP, creates an egress SNAT rule to convert source ip from PodIp to pod's GlobalIp.

Reviewer 2: ... pod's GlobalIp -> ...pod's GlobalIp on the GatewayNode?

> * Whenever a service is annotated with Global IP, creates an ingress rule to direct all traffic destined to service's GlobalIp to the service's `kube-proxy` IPTables chain which in turn directs traffic to ServiceIp.

Reviewer 2: ...in turn directs traffic to ServiceIp -> ... in turn directs traffic to backend pods.

* On deletion of pod/service, clean up the rules from the gateway node.

> Since this implementation currently relies on KubeProxy, globalnet only works with deployments that use `kube-proxy`

Reviewer: ```suggestion
Globalnet currently relies on `kube-proxy` and thus will only work with deployments that use `kube-proxy`.
```

<!-- TODO: Add a block diagram showing the solution -->

## Service Discovery - Lighthouse

> Connectivity is only part of solution, pods still need to know the IPs of services on remote clusters.

Reviewer: ```suggestion
Connectivity is only part of the solution as pods still need to know the IPs of services on remote clusters.
```

> This is achieved by enhancing [lighthouse](https://github.com/submariner-io/lighthouse) with support for GlobalNet. Lighthouse controller adds the service's GlobalIp to the `MultiClusterSerivce` object that is distributed to all clusters. The [lighthouse plugin](https://github.com/submariner-io/lighthouse/tree/master/plugin/lighthouse) then uses this Service's GlobalIp when replying to DNS queries for the service.

Reviewer 2: nit: MultiClusterSerivce -> MultiClusterService

> This is achieved by enhancing [lighthouse](https://github.com/submariner-io/lighthouse) with support for GlobalNet. Lighthouse controller adds the service's GlobalIp to the `MultiClusterSerivce` object that is distributed to all clusters. The [lighthouse plugin](https://github.com/submariner-io/lighthouse/tree/master/plugin/lighthouse) then uses this Service's GlobalIp when replying to DNS queries for the service.

Reviewer: ```suggestion
This is achieved by enhancing [lighthouse](https://github.com/submariner-io/lighthouse) with support for GlobalNet. The Lighthouse controller adds the service's GlobalIp to the `MultiClusterSerivce` object that is distributed to all clusters. The [lighthouse plugin](https://github.com/submariner-io/lighthouse/tree/master/plugin/lighthouse) then uses the service's GlobalIp when replying to DNS queries for the service.
```

# Installation

TBD.

# Testing

Refer [Testing guide](testing.md) for instructions on how to test, same steps apply.

# Known Issues

# Building

> Nothing extra needs to be done to build `submariner-globalnet`. Standard submariner build will now build `submariner-globalnet` binary too.

Reviewer: ```suggestion
Nothing extra needs to be done to build `submariner-globalnet` as it is built with the standard submariner build.
```

# TODO

> Not everything mentioned in architecture is currently implemented. This section lists work items that are pending.

Reviewer: ```suggestion
Not everything mentioned in the architecture is currently implemented. This section lists work items that are pending.
```

# Architecture

To support overlapping CIDRs in clusters connected through submariner, we're introducing a new component called Global Private Network, GlobalNet (`globalnet`). This GlobalNet is a virtual network specifically to support submariner's multi-cluster solution with a Global CIDR. Each cluster is given a subnet from this Global Private Network, configured as new cluster parameter `GlobalCIDR` (e.g. 169.254.0.0/16) which is configurable at time of deployment

Once configured, each service and pod that requires cross-cluster access is allocated an IP, `GlobalIp` from this `GlobalCIDR` and it is annotated on the Pod/Service object. This `GlobalIp` is used for all cross-cluster communication to and from to this Pod/Service. Routing and IPTable/OVS/OVN rules are configured to use this IP for ingress and egress. All address translations occur on the Gateway node.

## submariner-globalnet

> Submariner GlobalNet is a coponent that provides cross-cluster connectivity between pods and services using their GlobalIps. Compiled as binary `submariner-globalnet`, it is the controller that is responsible for maintaining a pool of global IPs, allocating IPs from the globalIp pool to pods and services, annotating services and pods with their global Ip, and configuring the required rules on gateway node to provide cross-cluster connectivity using GlobalIps. It mainly consists of two key components: the IP Address Manager and Globalnet.

Reviewer 2: nit: coponent -> component
